libcudacxx/include/cuda/std/detail/libcxx/include/__utility/pair.h

```diff
@@ -69,707 +70,704 @@
-    pair(piecewise_construct_t __pc,
-         tuple<_Args1...> __first_args, tuple<_Args2...> __second_args)
-        noexcept((is_nothrow_constructible<first_type, _Args1...>::value &&
-                    is_nothrow_constructible<second_type, _Args2...>::value))
-        : pair(__pc, __first_args, __second_args,
-                __make_tuple_indices_t<sizeof...(_Args1)>(),
-                __make_tuple_indices_t<sizeof...(_Args2)>()) {}
-
-    _LIBCUDACXX_INLINE_VISIBILITY _LIBCUDACXX_CONSTEXPR_AFTER_CXX17
-    pair& operator=(__conditional_t<
-                        is_copy_assignable<first_type>::value &&
-                        is_copy_assignable<second_type>::value,
-                    pair, __nat> const& __p)
-        noexcept(is_nothrow_copy_assignable<first_type>::value &&
-                   is_nothrow_copy_assignable<second_type>::value)
-    {
-        first = __p.first;
-        second = __p.second;
-        return *this;
-    }
+// We need to synthesize the copy / move assignment if it would be implicitly deleted as a member of a class
+// In that case _T1 would be copy assignable but _TestSynthesizeAssignment<_T1> would not
+// This happens e.g for reference types
+template <class _T1>
+struct _TestSynthesizeAssignment
+{
+  _T1 __dummy;
+};
 
-#if defined(__cuda_std__) && !defined(__CUDACC_RTC__)
-    template<class _UT1 = _T1, __enable_if_t<is_copy_assignable<_UT1>::value &&
-                                             is_copy_assignable<_T2>::value, int> = 0>
-    _LIBCUDACXX_HOST _LIBCUDACXX_CONSTEXPR_AFTER_CXX17
-    pair& operator=(::std::pair<_T1, _T2> const& __p)
-        noexcept(is_nothrow_copy_assignable<first_type>::value &&
-                   is_nothrow_copy_assignable<second_type>::value)
-    {
-        first = __p.first;
-        second = __p.second;
-        return *this;
-    }
-#endif // defined(__cuda_std__) && !defined(__CUDACC_RTC__)
+template <class _T1, class _T2>
+struct __must_synthesize_assignment
+    : integral_constant<bool,
+                        (_LIBCUDACXX_TRAIT(is_copy_assignable, _T1) && _LIBCUDACXX_TRAIT(is_copy_assignable, _T2)
+                         && !(_LIBCUDACXX_TRAIT(is_copy_assignable, _TestSynthesizeAssignment<_T1>)
+                              && _LIBCUDACXX_TRAIT(is_copy_assignable, _TestSynthesizeAssignment<_T2>)))
+                          || (_LIBCUDACXX_TRAIT(is_copy_assignable, _T1) && _LIBCUDACXX_TRAIT(is_copy_assignable, _T2)
+                              && !(_LIBCUDACXX_TRAIT(is_copy_assignable, _TestSynthesizeAssignment<_T1>)
+                                   && _LIBCUDACXX_TRAIT(is_copy_assignable, _TestSynthesizeAssignment<_T2>)))>
+{};
+
+// base class to ensure `is_trivially_copyable` when possible
+template <class _T1, class _T2, bool = __must_synthesize_assignment<_T1, _T2>::value>
+struct __pair_base
+{
+  _T1 first;
+  _T2 second;
+
+  template <class _Constraints                                                 = __pair_constraints<_T1, _T2>,
```

Comment:
Side note: I *hate* how this gets formatted.

Reply:
> but the alternative is worse

@miscco I'm not sure. I'd gladly avoid aligning by operators altogether.

Reply:
The issue is that usually aligning template parameters is highly beneficial and improves readability a lot. 

Its only when you have larger defaults like here that it breaks down